crates/api/src/trap.rs

```diff
@@ -1,11 +1,88 @@
+use crate::instance::Instance;
+use crate::r#ref::HostRef;
+use std::fmt;
+use std::sync::{Arc, Mutex};
 use thiserror::Error;
 
+#[derive(Debug)]
+pub struct FrameInfo;
+
+impl FrameInfo {
+    pub fn instance(&self) -> *const Instance {
+        unimplemented!("FrameInfo::instance");
+    }
+
+    pub fn func_index() -> usize {
+        unimplemented!("FrameInfo::func_index");
+    }
+
+    pub fn func_offset() -> usize {
+        unimplemented!("FrameInfo::func_offset");
+    }
+
+    pub fn module_offset() -> usize {
+        unimplemented!("FrameInfo::module_offset");
+    }
+}
+
+#[derive(Debug)]
+pub struct TrapInfo {
+    message: String,
+    trace: Vec<FrameInfo>,
+}
+
+impl TrapInfo {
+    pub fn new<I: Into<String>>(message: I) -> Self {
+        Self {
+            message: message.into(),
+            trace: vec![],
+        }
+    }
+
+    /// Returns a reference the `message` stored in `Trap`.
+    pub fn message(&self) -> &str {
+        &self.message
+    }
+
+    pub fn origin(&self) -> Option<&FrameInfo> {
+        self.trace.first()
+    }
+
+    pub fn trace(&self) -> &[FrameInfo] {
+        &self.trace
+    }
+}
+
+// A struct to hold unsafe TrapInfo host reference, designed
+// to be Send-able. The only access for it provided via the
+// Trap::trap_info_unchecked() method.
+struct UnsafeTrapInfo(HostRef<TrapInfo>);
+
+impl UnsafeTrapInfo {
+    fn trap_info(&self) -> HostRef<TrapInfo> {
+        self.0.clone()
+    }
+}
+
+unsafe impl Send for UnsafeTrapInfo {}
```

Comment:
Would it be possible to avoid this? Could we implement whatever the C API needs via some other mechanism (or just leave it unimplemented for the trap object?)

This is a super dangerous `unsafe impl` to have here and would be very difficult to explain to consumers how to use `trap_info_unchecked` safely

Reply:
I'm still working on mechanism to associate API objects with rc+hostinfo+finalizer at C API level. This PR just a step allowing Instance::new() to return a Trap object created by user code/callback.

Reply:
Ah ok, do you think this is worth landing in the meantime while the other mechanism is sorted out?

If so, mind making the unsafe method `pub(crate)` and adding a doc-block indicating that it's scheduled for deletion? (also a comment on this `unsafe impl` pointing to that method)